rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
+        Self {
+            handler,
+            subscriptions,
+            subscriptions_buffers: RefCell::new(heapless::Vec::new()),
+            buffers,
+        }
+    }
+
+    /// Answer a responding exchange using the `DataModelHandler` instance wrapped by this exchange handler.
+    pub async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let mut timeout_instant = None;
+
+        loop {
+            let mut repeat = false;
+
+            if exchange.rx().is_err() {
+                exchange.recv_fetch().await?;
+            }
+
+            let meta = exchange.rx()?.meta();
+            if meta.proto_id != PROTO_ID_INTERACTION_MODEL {
+                Err(ErrorCode::InvalidProto)?;
+            }
+
+            match meta.opcode::<OpCode>()? {
+                OpCode::ReadRequest => self.read(exchange).await?,
+                OpCode::WriteRequest => {
+                    repeat = self.write(exchange, timeout_instant.take()).await?;
+                }
+                OpCode::InvokeRequest => self.invoke(exchange, timeout_instant.take()).await?,
+                OpCode::SubscribeRequest => self.subscribe(exchange).await?,
+                OpCode::TimedRequest => {
+                    timeout_instant = Some(self.timed(exchange).await?);
+                    repeat = true;
+                }
+                opcode => {
+                    error!("Invalid opcode: {:?}", opcode);
+                    Err(ErrorCode::InvalidOpcode)?
+                }
+            }
 
-pub struct DataModel<T>(T);
+            if !repeat {
+                break;
+            }
+        }
 
-impl<T> DataModel<T> {
-    pub fn new(handler: T) -> Self {
-        Self(handler)
+        exchange.acknowledge().await?;
+        exchange.matter().notify_changed();
+
+        Ok(())
     }
 
-    pub async fn handle<'r, 'p>(
+    async fn read(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        let Some(mut tx) = self.tx_buffer(exchange).await? else {
+            return Ok(());
+        };
+
+        let mut wb = WriteBuf::new(&mut tx);
+
+        let metadata = self.handler.lock().await;
+
+        let req = ReadReq::from_tlv(&get_root_node_struct(exchange.rx()?.payload())?)?;
+        debug!("IM: Read request: {:?}", req);
+
+        let req = ReportDataReq::Read(&req);
+
+        let accessor = exchange.accessor()?;
+
+        let awaits = metadata.node().read(&req, None, &accessor).any(|item| {
+            item.map(|attr| self.handler.read_awaits(&attr))
+                .unwrap_or(false)
+        });
+
+        if !awaits {
+            let node = metadata.node();
+            let mut attrs = node.read(&req, None, &accessor).peekable();
+
+            if !req
+                .respond(&self.handler, None, &mut attrs, &mut wb, true)
+                .await?
+            {
+                drop(attrs);
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
```

Comment:
I see here we send a report_data with the writebuffer as a slice. Does the rust code currently support chunking? I.e. when a large read is requested (e.g. a "subscribe/read *" for a whildard of much data or overly large list reads), then the implementation is expected to send the response in multiple report-data elements each with a subset of data.

Is this handled anywhere? Trying to understand how chunking works (if it exists) or if chunking is not yet here and this sends the entire thing assuming we can always send a full WB (i.e. maybe respond already does chunking)

Reply:
I see some chunking code below. Closing this for me to read that bit of code a bit more.